build_tools/docker/manage_images.py

```diff
@@ -0,0 +1,249 @@
+      default='latest',
+      help='Tag for the images to build. Defaults to `latest` (which is good '
+      'for testing changes in a PR). Use `prod` to update the images that the '
+      'CI caches.')
+  parser.add_argument('--pull',
+                      action='store_true',
+                      help='Pull the specified image before building.')
+  parser.add_argument('--build',
+                      action='store_true',
+                      help='Build new images from the current Dockerfiles.')
+  parser.add_argument(
+      '--push',
+      action='store_true',
+      help='Push the built images to GCR. Requires gcloud authorization.')
+  parser.add_argument(
+      '--update_references',
+      action='store_true',
+      help='Update all references to the specified images to point at the new'
+      ' digest.')
+  parser.add_argument(
+      '--dry-run',
+      '-n',
+      action='store_true',
+      help='Print output without building or pushing any images.')
+
+  args = parser.parse_args()
+  for image in args.images:
+    if image == 'all':
+      # Sort for a determinstic order
+      args.images = sorted(IMAGES_TO_DEPENDENCIES.keys())
+    elif image not in IMAGES_TO_DEPENDENCIES:
+      raise parser.error('Expected --image to be one of:\n'
+                         f'  {IMAGES_HELP}\n'
+                         f'but got `{image}`.')
+  return args
+
+
+def get_ordered_images_to_process(images):
+  unmarked_images = list(images)
+  # Python doesn't have a builtin OrderedSet
+  marked_images = set()
+  order = []
+
+  def visit(image):
+    if image in marked_images:
+      return
+    for dependent_images in IMAGES_TO_DEPENDENT_IMAGES[image]:
+      visit(dependent_images)
+    marked_images.add(image)
+    order.append(image)
+
+  while unmarked_images:
+    visit(unmarked_images.pop())
+
+  order.reverse()
+  return order
+
+
+def stream_command(command, dry_run=False):
+  print(f'Running: `{" ".join(command)}`')
+  if dry_run:
+    return 0
+  process = subprocess.Popen(command,
+                             bufsize=1,
+                             stderr=subprocess.STDOUT,
+                             stdout=subprocess.PIPE,
+                             text=True)
+  for line in process.stdout:
+    print(line, end='')
+
+  if process.poll() is None:
+    raise RuntimeError('Unexpected end of output while process is not finished')
+  return process.poll()
+
+
+def check_stream_command(command, dry_run=False):
+  exit_code = stream_command(command, dry_run=dry_run)
+  if exit_code != 0:
+    print(f'Command failed with exit code {exit_code}: `{" ".join(command)}`')
+    sys.exit(exit_code)
+
+
+def get_repo_digest(image):
+  inspect_command = [
+      'docker',
+      'image',
+      'inspect',
+      f'{image}',
+      '-f',
+      '{{index .RepoDigests 0}}',
+  ]
+  inspect_process = subprocess.run(inspect_command,
+                                   text=True,
+                                   capture_output=True,
```

Comment:
These arguments were added in `python3.7` so this [won't run on `python3.6`](https://docs.python.org/3.6/library/subprocess.html).

Previously we were limiting ourselves to using `python3.6` because TensorFlow, but it looks like they got around to [supporting `python3.8`](https://www.tensorflow.org/install/pip). I tested and we can now build the integrations with `python3.8`, so I'm all for assuming in our scripts that people are at `python3.8+` (yay new features!). Thoughts?

@ScottTodd might have some thoughts on this too.

Reply:
Oh I didn't mean to require 3.7 or higher. I think sticking with 3.6 would be reasonable unless we've got some strong new feature we need. I'll take a look at making it compatible.

Reply:
`dataclasses`, guaranteed dictionary ordering, assignment operators and [self documenting f-strings](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging) are probably my favorite new features.